test_data/proxy/src/com/intellij/memory/agent/AgentExtractor.java

```diff
@@ -0,0 +1,82 @@
+package com.intellij.memory.agent;
+
+import java.io.*;
+import java.util.Locale;
+
+class AgentExtractor {
+    private static File file = null;
+
+    private AgentExtractor() {
+
+    }
+
+    static File extract(File directory) throws Exception {
+        if (file != null) {
+            return file;
+        }
+
+        AgentLibraryType agentType = detectAgentKind();
+        if (agentType == null) {
+            throw new Exception("Couldn't detect agent kind for current OS");
+        }
+
+        String agentFileName = agentType.prefix + "memory_agent";
+        File f = new File(String.format("%s%c%s%s", directory.getAbsolutePath(), File.separatorChar, agentFileName, agentType.suffix));
+        if (f.exists() && !f.isDirectory()) {
+            file = f;
+            return file;
+        }
+
+        file = File.createTempFile(agentFileName, agentType.suffix, directory);
+        try(OutputStream outputStream = new FileOutputStream(file);
+            InputStream inputStream = AgentExtractor.class.getClassLoader().getResourceAsStream( "bin/" + agentFileName + agentType.suffix)) {
+            if (inputStream == null) {
+                throw new MemoryAgentException("Agent binary library couldn't be found");
+            }
+
+            copyStream(inputStream, outputStream);
+            return file;
+        }
+    }
+
+    private static void copyStream(InputStream in, OutputStream out) throws IOException {
+        byte[] buffer = new byte[1024];
```

Comment:
Isn't the buffer too small? Also, using `BufferedInputStream` might be more clear (it seems to use an 8K buffer).